backend/llm/openai.py

```diff
@@ -2,136 +2,119 @@
-    question_generator: LLMChain = None
-    doc_chain: LLMChain = None
-    qa: ConversationalRetrievalChain = None
-
-    # Streaming
-    callback: AsyncIteratorCallbackHandler = None
     streaming: bool = False
 
-    class Config:
-        # Allowing arbitrary types for class validation
-        arbitrary_types_allowed = True
-
     def __init__(
         self,
-        model: str,
+        model_name: str,
         brain_id: str,
         temperature: float,
         chat_id: str,
         max_tokens: int,
         user_openai_api_key: str,
         streaming: bool = False,
-    ) -> "BrainPicking":
+    ):
         """
         Initialize the BrainPicking class by setting embeddings, supabase client, vector store, language model and chains.
-        :param model: Language model name to be used.
-        :param user_brain_idid: The brain id to be used for CustomSupabaseVectorStore.
-        :return: BrainPicking instance
+        :return: OpenAIBrainPicking instance
         """
         super().__init__(
-            model=model,
+            model_name=model_name,
             brain_id=brain_id,
             chat_id=chat_id,
             max_tokens=max_tokens,
             temperature=temperature,
             user_openai_api_key=user_openai_api_key,
+            streaming=streaming,
         )
-        # If user provided an API key, update the settings
-        if user_openai_api_key is not None:
-            self.settings.openai_api_key = user_openai_api_key
 
+        # Set the class attributes
         self.temperature = temperature
-        self.embeddings = OpenAIEmbeddings(openai_api_key=self.settings.openai_api_key)
-        self.supabase_client = create_client(
-            self.settings.supabase_url, self.settings.supabase_service_key
+        self.model_name = model_name
+        self.chat_id = chat_id
+        self.brain_id = brain_id
+        self.max_tokens = max_tokens
+
+    @property
+    def embeddings(self) -> OpenAIEmbeddings:
+        return OpenAIEmbeddings(openai_api_key=self.openai_api_key)
+
+    @property
+    def supabase_client(self) -> Client:
+        return create_client(
+            self.brain_settings.supabase_url, self.brain_settings.supabase_service_key
         )
-        self.llm_name = model
-        self.vector_store = CustomSupabaseVectorStore(
+
+    @property
+    def vector_store(self) -> CustomSupabaseVectorStore:
+        return CustomSupabaseVectorStore(
             self.supabase_client,
             self.embeddings,
             table_name="vectors",
-            brain_id=brain_id,
+            brain_id=self.brain_id,
         )
 
-        self.question_llm = self._create_llm(
-            model_name=self.llm_name,
-            streaming=False,
-        )
-        self.question_generator = LLMChain(
-            llm=self.question_llm, prompt=CONDENSE_QUESTION_PROMPT
-        )
+    @property
+    def question_llm(self) -> LLM:
+        return self._create_llm(streaming=False)
```

Comment:
Why is streaming always set as 'False', not using `self.streaming` ?

Reply:
You only want streaming on the doc_chain llm as that is the final output. The question_chain llm is the intermediary llm model.